src/dbnode/storage/index/results.go

```diff
@@ -178,11 +182,27 @@ func (r *results) TotalDocsCount() int {
 }
 
 func (r *results) Finalize() {
+	r.Lock()
+
+	if r.totalDocsCount < r.maxObservedSize {
+		r.underutilizedRunLength++
+	} else {
+		r.maxObservedSize = r.totalDocsCount
+		r.underutilizedRunLength = 0
+	}
+
+	// If the size of reusable result object is not fully utilized for a long enough time,
+	// we want to drop it from the pool and prevent holding on to excessive memory indefinitely.
+	putBackToPool := r.underutilizedRunLength < 100
```

Comment:
Nice I like this approach, where we should purge pooled results sizes unless they are common